From 4ccc753e4f957e62637a59f53f593a79cd31ab05 Mon Sep 17 00:00:00 2001
From: Kenton Varda <kenton@cloudflare.com>
Date: Sat, 2 Mar 2024 09:00:18 -0600
Subject: Add ValueSerializer::SetTreatFunctionsAsHostObjects().

Previously, ValueSerializer would always refuse to serialize functions. This commit gives the embedder the option to handle them as host objects.

This is intended for use in an RPC system, where a function can be "serialized" by replacing it with a stub which, when invoked, performs an RPC back to the originating isolate in order to execute the original function there.

diff --git a/include/v8-value-serializer.h b/include/v8-value-serializer.h
index 596be18adeb3a5a81794aaa44b1d347dec6c0c7d..141f138e08de849e3e02b3b2b346e643b9e40c70 100644
--- a/include/v8-value-serializer.h
+++ b/include/v8-value-serializer.h
@@ -195,6 +195,15 @@ class V8_EXPORT ValueSerializer {
    */
   void SetTreatArrayBufferViewsAsHostObjects(bool mode);
 
+  /**
+   * Indicate whether to treat Functions as host objects,
+   * i.e. pass them to Delegate::WriteHostObject. This should not be
+   * called when no Delegate was passed.
+   *
+   * The default is not to treat Functions as host objects.
+   */
+  void SetTreatFunctionsAsHostObjects(bool mode);
+
   /**
    * Write raw data in various common formats to the buffer.
    * Note that integer types are written in base-128 varint format, not with a
diff --git a/src/api/api.cc b/src/api/api.cc
index a0988b509630a854c167cb634dde5b059ec9c0c1..9954e6a9182d1908b2e92150fba14c31dc23455f 100644
--- a/src/api/api.cc
+++ b/src/api/api.cc
@@ -3448,6 +3448,10 @@ void ValueSerializer::SetTreatArrayBufferViewsAsHostObjects(bool mode) {
   private_->serializer.SetTreatArrayBufferViewsAsHostObjects(mode);
 }
 
+void ValueSerializer::SetTreatFunctionsAsHostObjects(bool mode) {
+  private_->serializer.SetTreatFunctionsAsHostObjects(mode);
+}
+
 Maybe<bool> ValueSerializer::WriteValue(Local<Context> context,
                                         Local<Value> value) {
   auto i_isolate = reinterpret_cast<i::Isolate*>(context->GetIsolate());
diff --git a/src/objects/value-serializer.cc b/src/objects/value-serializer.cc
index 2e574704650c3b5d860fc21d083e9e6fed9ddfcf..0c82ae8299f2f45bee34177c3d1b663cede0c02e 100644
--- a/src/objects/value-serializer.cc
+++ b/src/objects/value-serializer.cc
@@ -328,6 +328,10 @@ void ValueSerializer::SetTreatArrayBufferViewsAsHostObjects(bool mode) {
   treat_array_buffer_views_as_host_objects_ = mode;
 }
 
+void ValueSerializer::SetTreatFunctionsAsHostObjects(bool mode) {
+  treat_functions_as_host_objects_ = mode;
+}
+
 void ValueSerializer::WriteTag(SerializationTag tag) {
   uint8_t raw_tag = static_cast<uint8_t>(tag);
   WriteRawBytes(&raw_tag, sizeof(raw_tag));
@@ -596,8 +600,13 @@ Maybe<bool> ValueSerializer::WriteJSReceiver(Handle<JSReceiver> receiver) {
 
   // Eliminate callable and exotic objects, which should not be serialized.
   InstanceType instance_type = receiver->map()->instance_type();
-  if (IsCallable(*receiver) || (IsSpecialReceiverInstanceType(instance_type) &&
-                                instance_type != JS_SPECIAL_API_OBJECT_TYPE)) {
+  if (IsCallable(*receiver)) {
+    if (treat_functions_as_host_objects_) {
+      return WriteHostObject(Handle<JSObject>::cast(receiver));
+    }
+    return ThrowDataCloneError(MessageTemplate::kDataCloneError, receiver);
+  } else if (IsSpecialReceiverInstanceType(instance_type) &&
+             instance_type != JS_SPECIAL_API_OBJECT_TYPE) {
     return ThrowDataCloneError(MessageTemplate::kDataCloneError, receiver);
   }
 
diff --git a/src/objects/value-serializer.h b/src/objects/value-serializer.h
index c3a44e9f64f944641eea417584429a0b147ef654..fcf6e66d44508c2c29889354f3794d37cc22b52d 100644
--- a/src/objects/value-serializer.h
+++ b/src/objects/value-serializer.h
@@ -102,6 +102,15 @@ class ValueSerializer {
    */
   void SetTreatArrayBufferViewsAsHostObjects(bool mode);
 
+  /*
+   * Indicate whether to treat Functions as host objects,
+   * i.e. pass them to Delegate::WriteHostObject. This should not be
+   * called when no Delegate was passed.
+   *
+   * The default is not to treat Functions as host objects.
+   */
+  void SetTreatFunctionsAsHostObjects(bool mode);
+
  private:
   // Managing allocations of the internal buffer.
   Maybe<bool> ExpandBuffer(size_t required_capacity);
@@ -182,6 +191,7 @@ class ValueSerializer {
   size_t buffer_capacity_ = 0;
   bool has_custom_host_objects_ = false;
   bool treat_array_buffer_views_as_host_objects_ = false;
+  bool treat_functions_as_host_objects_ = false;
   bool out_of_memory_ = false;
   Zone zone_;
   uint32_t version_;
